src/NuGet.Core/NuGet.Packaging/Signing/SignatureUtility.cs

```diff
@@ -0,0 +1,487 @@
+            {
+                throw new SignatureException(errors.InvalidSignature, Strings.SigningCertificateInvalid);
+            }
+
+            if (!IsMatch(signerCertificate, signingCertificate.Certificates[0]))
+            {
+                throw new SignatureException(errors.InvalidSignature, Strings.SigningCertificateCertificateNotFound);
+            }
+
+            var chain = GetCertificateChain(signerCertificate, extraStore, includeChain);
+
+            if (chain == null || chain.Count == 0)
+            {
+                throw new SignatureException(errors.ChainBuildingFailed, Strings.CertificateChainBuildFailed);
+            }
+
+            return chain;
+        }
+
+        private static bool IsMatch(
+            X509Certificate2 certificate,
+            EssCertIdV2 essCertIdV2,
+            Errors errors,
+            bool isIssuerSerialRequired)
+        {
+            if (isIssuerSerialRequired)
+            {
+                if (essCertIdV2.IssuerSerial == null ||
+                    essCertIdV2.IssuerSerial.GeneralNames.Count == 0)
+                {
+                    throw new SignatureException(errors.InvalidSignature, errors.InvalidSignatureString);
+                }
+            }
+
+            if (essCertIdV2.IssuerSerial != null)
+            {
+                if (!AreSerialNumbersEqual(essCertIdV2.IssuerSerial, certificate))
+                {
+                    return false;
+                }
+
+                var generalName = essCertIdV2.IssuerSerial.GeneralNames.FirstOrDefault();
+
+                if (generalName != null &&
+                    generalName.DirectoryName != null &&
+                    generalName.DirectoryName.Name != certificate.IssuerName.Name)
+                {
+                    return false;
+                }
+            }
+
+            var hashAlgorithmName = CryptoHashUtility.OidToHashAlgorithmName(essCertIdV2.HashAlgorithm.Algorithm.Value);
+            var actualHash = CertificateUtility.GetHash(certificate, hashAlgorithmName);
+
+            return essCertIdV2.CertificateHash.SequenceEqual(actualHash);
+        }
+
+        private static bool IsMatch(X509Certificate2 certificate, EssCertId essCertId)
+        {
+            if (essCertId.IssuerSerial != null)
+            {
+                if (!AreSerialNumbersEqual(essCertId.IssuerSerial, certificate))
+                {
+                    return false;
+                }
+
+                var generalName = essCertId.IssuerSerial.GeneralNames.FirstOrDefault();
+
+                if (generalName != null &&
+                    generalName.DirectoryName != null &&
+                    generalName.DirectoryName.Name != certificate.IssuerName.Name)
+                {
+                    return false;
+                }
+            }
+
+            byte[] actualHash;
+
+            using (var hashAlgorithm = CryptoHashUtility.GetSha1HashProvider())
```

Comment:
Just curious, why isn't this part of `CertificateUtility.GetHash` and `HashAlgorithmName`?

Reply:
`HashAlgorithmName` is used all over the place, for validating hash algorithm choice, generating hashes for signing and for validation.  If I added `HashAlgorithmName.SHA1` I'd have to review every call site to ensure SHA-1 couldn't used.  The only place for SHA-1 is in validation of the `signing-certificate` attribute, where the hash algorithm is hardcoded (by specification) to SHA-1.